### PR TITLE
Replace thread_local with generator resume arguments in box_region.

### LINKED ISSUE #68922: Rewrite `box_region.rs` to use generator resume arguments
[`box_region.rs`](https://github.com/rust-lang/rust/blob/3d8778d767f0dde6fe2bc9459f21ead8e124d8cb/src/librustc_data_structures/box_region.rs) makes use of generators, and passes arguments into them via thread-local storage. Now that https://github.com/rust-lang/rust/pull/68524 has landed, this can be rewritten to use generator resume arguments instead, which is much cleaner and possibly also faster.

### PR DESCRIPTION
**Pull request by gizmondo**
Fixes #68922.

r? @ghost

### CODE REVIEW

**Comment by gizmondo** [comment]
This is a plain replacement. @jonas-schievink, did you have something more comprehensive in mind?

**Comment by jonas-schievink** [comment]
Seems like a good start, but to be honest, I have no idea what this code is supposed to do (other than that it can now use resume arguments to do it).

r? @Zoxc

File: src/librustc_interface/passes.rs [hunk 1]
```diff
@@ -110,7 +110,8 @@ pub fn configure_and_expand(
     // its contents but the results of name resolution on those contents. Hopefully we'll push
     // this back at some point.
     let crate_name = crate_name.to_string();
-    let (result, resolver) = BoxedResolver::new(static move || {
+    let (result, resolver) = BoxedResolver::new(static move |mut action| {
+        let _ = action;
```
Comment (gizmondo) [inline src/librustc_interface/passes.rs]: that's just to suppress the warning, probably should be a better way...

**Comment by bjorn3** [comment]
`box_region` is only used in `librustc_interface`, so it is not performance critical. https://github.com/rust-lang/rust/blob/75ff3110ac6d8a0259023b83fd20d7ab295f8dd6/src/librustc_interface/passes.rs#L88-L92

**Comment by gizmondo** [comment]
Btw there was a [discussion](https://internals.rust-lang.org/t/making-the-compiler-interface-on-demand-driven/9099/5) that resume arguments can allow to get rid of unsafety in this module. 
But I haven't figured out a way to do it. I think for that to be possible _resume_ should be able to borrow its arguments for the duration of a single call. I.e. something like that should work:
```rust    
let mut gen = Box::pin(static move |mut arg: &mut i32| {
    loop {
        *arg += 1;
        arg = yield;
    }
});
let mut a = 1;
Pin::new(&mut gen).resume(&mut a);
```

Currently it doesn't compile. I'm not sure if I'm doing something wrong or it's just not supported yet.

**Comment by petrochenkov** [comment]
Since you are working on this, could you perhaps add some docs explaining what the "box region" does and why is it necessary? Right now it's entirely unclear.

**Comment by gizmondo** [comment]
I'll try.

**Comment by jonas-schievink** [comment]
@gizmondo for that to work you'd need https://github.com/rust-lang/rust/issues/68923, which is unfortunately very hard to solve

File: src/librustc_data_structures/box_region.rs [hunk 2]
```diff
@@ -152,10 +146,11 @@ macro_rules! box_region_allow_access {
                         let marker = $crate::box_region::Marker::<
                             for<$($lifetimes)*> fn(($($args,)*))
                         >::new();
-                        yield $crate::box_region::YieldType::Accessor(marker)
+                        $action = yield $crate::box_region::YieldType::Accessor(marker);
                     };
                 }
                 $crate::box_region::Action::Complete => break,
+                $crate::box_region::Action::Initial => panic!(),
```
Comment (jonas-schievink) [inline src/librustc_data_structures/box_region.rs:153]:
```suggestion
                $crate::box_region::Action::Initial => bug!("unexpected box_region action: {:?}", $action),
```

**Review by jonas-schievink — CHANGES_REQUESTED** [review]
Okay, looks basically good to me @gizmondo!

If you want to add some more comments to this module, feel free. Otherwise r=me with the more detailed panic message below.

**Comment by joelpalmer** [comment]
Ping from Triage: Closed due to inactivity. Please re-open with updates. Thanks for the PR! @gizmondo

**Pull request closed without merging**